File: internal/model/self_order.go
package model

import "time"

// SelfOrder 自营履约单据（承接 OrderCore self_ship，对接仓储扣库）。
type SelfOrder struct {
	ID            uint64     `gorm:"primaryKey" json:"id"`
	TenantID      uint64     `gorm:"index;not null" json:"tenantId"`
	SoNo          string     `gorm:"size:32;not null" json:"soNo"`
	Status        string     `gorm:"size:32;not null;default:ordered" json:"status"`
	WarehouseID   uint64     `gorm:"index;default:0" json:"warehouseId"`
	RefSoID       uint64     `gorm:"index;default:0" json:"refSoId"`
	RefTraceID    string     `gorm:"size:64;index" json:"refTraceId"`
	SaleAmount    float64    `gorm:"type:decimal(14,2);not null;default:0" json:"saleAmount"`
	CostAmount    float64    `gorm:"type:decimal(14,2);not null;default:0" json:"costAmount"`
	BuyerName     string     `gorm:"size:64" json:"buyerName"`
	BuyerPhone    string     `gorm:"size:32" json:"buyerPhone"`
	Address       string     `gorm:"type:text" json:"address"`
	Remark        string     `gorm:"type:text" json:"remark"` // 系统备注（如 OMS自营 / 取消原因）
	SourceChannel string     `gorm:"size:32;index" json:"sourceChannel"`
	Platform      string     `gorm:"size:32" json:"platform"`
	ShopName      string     `gorm:"size:128" json:"shopName"`
	// ManualSourceName 手工单订单来源名称（来自 OrderCore 手工订单来源字典）
	ManualSourceName string `gorm:"size:128" json:"manualSourceName"`
	BuyerRemark   string     `gorm:"type:text" json:"buyerRemark"`
	SellerRemark  string     `gorm:"type:text" json:"sellerRemark"`
	FenFaRemark   string     `gorm:"type:text" json:"fenFaRemark"`
	PrinterRemark string     `gorm:"type:text" json:"printerRemark"`
	StockDeducted bool       `gorm:"default:false;index" json:"stockDeducted"`
	StockError    string     `gorm:"type:text" json:"stockError"`
	// PayStatus 付款进度：unpaid|partial|paid（对齐分销/采购）
	PayStatus string `gorm:"size:16;default:unpaid;index" json:"payStatus"`
	// PaidAt 最早一笔已付时间（有付款记录时填写）
	PaidAt      *time.Time `json:"paidAt,omitempty"`
	OrderedAt   *time.Time `json:"orderedAt"`
	ShippedAt   *time.Time `json:"shippedAt"`
	CompletedAt *time.Time `json:"completedAt"`
	CreatedAt   time.Time  `json:"createdAt"`
	UpdatedAt   time.Time  `json:"updatedAt"`
	Items       []SelfOrderItem `gorm:"foreignKey:SelfOrderID" json:"items,omitempty"`
}

// SelfPayment 自营单付款记录。
type SelfPayment struct {
	ID           uint64     `gorm:"primaryKey" json:"id"`
	TenantID     uint64     `gorm:"index;not null" json:"tenantId"`
	SelfOrderID  uint64     `gorm:"index;not null" json:"selfOrderId"`
	PayAmount    float64    `gorm:"type:decimal(14,2);not null" json:"payAmount"`
	PayMethod    string     `gorm:"size:32" json:"payMethod"`
	PayAccount   string     `gorm:"size:128" json:"payAccount"`
	PayeeAccount string     `gorm:"size:128" json:"payeeAccount"`
	PayeeName    string     `gorm:"size:128" json:"payeeName"`
	PayStatus    string     `gorm:"size:16;default:paid" json:"payStatus"`
	PaidAt       *time.Time `json:"paidAt"`
	Remark       string     `gorm:"type:text" json:"remark"`
	CreatedAt    time.Time  `json:"createdAt"`
	UpdatedAt    time.Time  `json:"updatedAt"`
}

func (SelfPayment) TableName() string { return "self_payments" }

func (SelfOrder) TableName() string { return "self_orders" }

type SelfOrderItem struct {
	ID            uint64    `gorm:"primaryKey" json:"id"`
	TenantID      uint64    `gorm:"index;not null" json:"tenantId"`
	SelfOrderID   uint64    `gorm:"index;not null" json:"selfOrderId"`
	PimSkuID      uint64    `gorm:"index;default:0" json:"pimSkuId"`
	SkuCode       string    `gorm:"size:64" json:"skuCode"`
	ProductName   string    `gorm:"size:512" json:"productName"`
	SkuSpecs      string    `gorm:"size:256" json:"skuSpecs"`
	PicURL        string    `gorm:"size:512" json:"picUrl"`
	Qty           int       `gorm:"not null;default:1" json:"qty"`
	SaleUnitPrice float64   `gorm:"type:decimal(12,2);not null;default:0" json:"saleUnitPrice"`
	SaleAmount    float64   `gorm:"type:decimal(14,2);not null;default:0" json:"saleAmount"`
	InvSkuID      uint64    `gorm:"index;default:0" json:"invSkuId"`
	InvSkuCode    string    `gorm:"size:64" json:"invSkuCode"`
	CostUnitPrice float64   `gorm:"type:decimal(12,2);not null;default:0" json:"costUnitPrice"`
	CostAmount    float64   `gorm:"type:decimal(14,2);not null;default:0" json:"costAmount"`
	RefSoID        uint64    `gorm:"index;default:0" json:"refSoId"`
	RefOrderItemID uint64    `gorm:"index;default:0" json:"refOrderItemId"`
	RefOrderNo     string    `gorm:"size:64;index" json:"refOrderNo"`
	Remark         string    `gorm:"type:text" json:"remark"`
	CreatedAt      time.Time `json:"createdAt"`
	UpdatedAt      time.Time `json:"updatedAt"`
}

func (SelfOrderItem) TableName() string { return "self_order_items" }

type SelfShipment struct {
	ID                  uint64     `gorm:"primaryKey" json:"id"`
	TenantID            uint64     `gorm:"index;not null" json:"tenantId"`
	SelfOrderID         uint64     `gorm:"index;not null" json:"selfOrderId"`
	ShipmentNo          string     `gorm:"size:32;not null" json:"shipmentNo"`
	Status              string     `gorm:"size:32;not null;default:shipped" json:"status"`
	CarrierCode         string     `gorm:"size:32" json:"carrierCode"`
	CarrierName         string     `gorm:"size:64" json:"carrierName"`
	TrackingNo          string     `gorm:"size:64" json:"trackingNo"`
	ShippedAt           *time.Time `json:"shippedAt"`
	ExpectedArrivalDate *time.Time `json:"expectedArrivalDate"`
	DeliveredAt         *time.Time `json:"deliveredAt"`
	ReceiverName        string     `gorm:"size:64" json:"receiverName"`
	ReceiverPhone       string     `gorm:"size:32" json:"receiverPhone"`
	ReceiverAddress     string     `gorm:"size:255" json:"receiverAddress"`
	CallbackOK          bool       `gorm:"default:false" json:"callbackOk"`
	StockDeducted       bool       `gorm:"default:false" json:"stockDeducted"`
	Remark              string     `gorm:"type:text" json:"remark"`
	CreatedAt           time.Time  `json:"createdAt"`
	UpdatedAt           time.Time  `json:"updatedAt"`
	Items               []SelfShipmentItem `gorm:"foreignKey:ShipmentID" json:"items,omitempty"`
}

func (SelfShipment) TableName() string { return "self_shipments" }

type SelfShipmentItem struct {
	ID              uint64    `gorm:"primaryKey" json:"id"`
	TenantID        uint64    `gorm:"index;not null" json:"tenantId"`
	ShipmentID      uint64    `gorm:"index;not null" json:"shipmentId"`
	SelfOrderItemID uint64    `gorm:"index;not null" json:"selfOrderItemId"`
	Qty             int       `gorm:"not null" json:"qty"`
	CreatedAt       time.Time `json:"createdAt"`
	UpdatedAt       time.Time `json:"updatedAt"`
}

func (SelfShipmentItem) TableName() string { return "self_shipment_items" }

type SelfAttachment struct {
	ID          uint64    `gorm:"primaryKey" json:"id"`
	TenantID    uint64    `gorm:"index;not null" json:"tenantId"`
	SelfOrderID uint64    `gorm:"index;not null" json:"selfOrderId"`
	PaymentID   uint64    `gorm:"index" json:"paymentId"`
	ShipmentID  uint64    `gorm:"index" json:"shipmentId"`
	FileType    string    `gorm:"size:32;not null" json:"fileType"`
	FileName    string    `gorm:"size:255;not null" json:"fileName"`
	FileURL     string    `gorm:"size:512;not null" json:"fileUrl"`
	UploadedBy  uint64    `json:"uploadedBy"`
	Remark      string    `gorm:"type:text" json:"remark"`
	CreatedAt   time.Time `json:"createdAt"`
	UpdatedAt   time.Time `json:"updatedAt"`
}

func (SelfAttachment) TableName() string { return "self_attachments" }

File: web/src/api/selfOrder.ts
import client, { unwrap, type PageData } from './client'

export interface SelfOrderListItem {
  id: number
  soNo: string
  status: string
  warehouseId: number
  refSoId: number
  refTraceId: string
  saleAmount: number
  costAmount: number
  payStatus?: string
  paidAt?: string
  sourceChannel?: string
  platform?: string
  shopName?: string
  buyerRemark?: string
  sellerRemark?: string
  fenFaRemark?: string
  printerRemark?: string
  skuSpecs?: string
  itemCount: number
  stockDeducted: boolean
  stockError: string
  orderedAt?: string
  shippedAt?: string
  createdAt: string
}

export interface SelfOrderItem {
  id: number
  pimSkuId: number
  skuCode: string
  productName: string
  skuSpecs: string
  picUrl: string
  qty: number
  saleUnitPrice: number
  saleAmount: number
  invSkuId: number
  invSkuCode: string
  costUnitPrice: number
  costAmount: number
  refSoId: number
  refOrderNo: string
  remark: string
}

export interface SelfOrderDetail {
  id: number
  soNo: string
  status: string
  warehouseId: number
  refSoId: number
  refTraceId: string
  saleAmount: number
  costAmount: number
  payStatus?: string
  paidAt?: string
  buyerName: string
  buyerPhone: string
  address: string
  remark: string
  sourceChannel?: string
  platform?: string
  shopName?: string
  buyerRemark?: string
  sellerRemark?: string
  fenFaRemark?: string
  printerRemark?: string
  stockDeducted: boolean
  stockError: string
  orderedAt?: string
  shippedAt?: string
  completedAt?: string
  createdAt: string
  updatedAt: string
  items: SelfOrderItem[]
}

export interface SelfShipment {
  id: number
  selfOrderId?: number
  shipmentNo: string
  status: string
  carrierCode?: string
  carrierName: string
  trackingNo: string
  shippedAt?: string
  expectedArrivalDate?: string
  deliveredAt?: string
  callbackOk: boolean
  stockDeducted?: boolean
  receiverName: string
  receiverPhone: string
  receiverAddress: string
  remark: string
  items?: { id?: number; selfOrderItemId: number; qty: number }[]
  createdAt?: string
}

export interface BindInvSkuInput {
  invSkuId: number
  invSkuCode?: string
  costUnitPrice?: number
}

export interface SelfShipInput {
  expressCompany: string
  expressNo: string
  remark?: string
  callback?: boolean
}

export interface WarehouseSku {
  id: number
  skuCode: string
  name: string
  pickName?: string
  lastPurchasePrice: number
}

export interface Warehouse {
  id: number
  code: string
  name: string
  status: number
  isDefault: number
}

/** 单据状态展示：已下单 / 已完成 / 已取消（内部 paid/发货中等均归为已下单） */
export const SELF_ORDER_STATUS_MAP: Record<
  string,
  { label: string; type: '' | 'success' | 'warning' | 'info' | 'danger' }
> = {
  draft: { label: '已下单', type: 'warning' },
  ordered: { label: '已下单', type: 'warning' },
  confirmed: { label: '已下单', type: 'warning' },
  paid: { label: '已下单', type: 'warning' },
  partial_shipped: { label: '已下单', type: 'warning' },
  shipped: { label: '已下单', type: 'warning' },
  completed: { label: '已完成', type: 'success' },
  cancelled: { label: '已取消', type: 'info' },
}

/** 单据状态筛选项 */
export const SELF_ORDER_DOC_STATUS_OPTIONS = [
  { value: 'ordered', label: '已下单' },
  { value: 'completed', label: '已完成' },
  { value: 'cancelled', label: '已取消' },
] as const

/** 发货状态：待发货 / 部分发货 / 已发货 */
export const SELF_SHIP_STATUS_MAP: Record<
  string,
  { label: string; type: '' | 'success' | 'warning' | 'info' | 'danger' }
> = {
  wait_ship: { label: '待发货', type: 'warning' },
  partial_shipped: { label: '部分发货', type: 'warning' },
  shipped: { label: '已发货', type: 'success' },
}

/** 付款状态筛选项 */
export const SELF_PAY_STATUS_OPTIONS = [
  { value: 'unpaid', label: '未付款' },
  { value: 'partial', label: '部分付款' },
  { value: 'paid', label: '已付清' },
] as const

export function deriveSelfDocStatus(status?: string): string {
  const s = (status || '').trim()
  if (s === 'completed') return 'completed'
  if (s === 'cancelled') return 'cancelled'
  if (!s) return ''
  return 'ordered'
}

export function deriveSelfShipStatus(status?: string): string {
  switch ((status || '').trim()) {
    case 'partial_shipped':
      return 'partial_shipped'
    case 'shipped':
    case 'completed':
      return 'shipped'
    case 'draft':
    case 'ordered':
    case 'paid':
    case 'confirmed':
      return 'wait_ship'
    default:
      return ''
  }
}

export async function listSelfOrders(params: {
  status?: string
  statuses?: string
  excludeStatuses?: string
  payStatus?: string
  shipStatus?: string
  refSoId?: number
  keyword?: string
  createdAtStart?: string
  createdAtEnd?: string
  orderedAtStart?: string
  orderedAtEnd?: string
  shippedAtStart?: string
  shippedAtEnd?: string
  page?: number
  pageSize?: number
} = {}) {
  const res = await client.get('/self-orders', { params })
  return unwrap<PageData<SelfOrderListItem>>(res)
}

export async function getSelfOrder(id: number) {
  return unwrap<SelfOrderDetail>(await client.get(`/self-orders/${id}`))
}

export async function shipSelfOrder(id: number, data: SelfShipInput) {
  return unwrap<SelfOrderDetail>(await client.post(`/self-orders/${id}/ship`, data))
}

export async function retryCallback(id: number, shipmentId?: number) {
  return unwrap<SelfOrderDetail>(
    await client.post(`/self-orders/${id}/retry-callback`, {}, {
      params: shipmentId ? { shipmentId } : undefined,
    }),
  )
}

export async function retryStock(id: number) {
  return unwrap<SelfOrderDetail>(await client.post(`/self-orders/${id}/retry-stock`))
}

export async function submitSelfOrder(id: number) {
  return unwrap<SelfOrderDetail>(await client.post(`/self-orders/${id}/submit`))
}

export async function markSelfOrderPaid(id: number) {
  return unwrap<SelfOrderDetail>(await client.post(`/self-orders/${id}/mark-paid`))
}

export async function completeSelfOrder(id: number) {
  return unwrap<SelfOrderDetail>(await client.post(`/self-orders/${id}/complete`))
}

export async function cancelSelfOrder(id: number) {
  return unwrap<SelfOrderDetail>(await client.post(`/self-orders/${id}/cancel`))
}

export async function deleteSelfOrder(id: number) {
  return unwrap(await client.delete(`/self-orders/${id}`))
}

export async function bindInvSku(itemId: number, data: BindInvSkuInput) {
  return unwrap<SelfOrderDetail>(await client.put(`/self-order-items/${itemId}/inv-sku`, data))
}

export async function updateItemCost(itemId: number, costUnitPrice: number) {
  return unwrap<SelfOrderDetail>(await client.put(`/self-order-items/${itemId}/cost`, { costUnitPrice }))
}

export async function listSelfShipments(selfOrderId: number) {
  return unwrap<SelfShipment[]>(await client.get(`/self-orders/${selfOrderId}/shipments`))
}

export async function searchWarehouseSkus(params: {
  keyword?: string
  page?: number
  pageSize?: number
} = {}) {
  const res = await client.get('/warehouse-skus/search', { params })
  return unwrap<PageData<WarehouseSku>>(res)
}

export async function listWarehouses(params: {
  keyword?: string
  page?: number
  pageSize?: number
} = {}) {
  const res = await client.get('/warehouses', { params })
  return unwrap<PageData<Warehouse>>(res)
}
